File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        def compare(left, right):
            for a, b in zip_longest(left, right, fillvalue=-1):
                if(a < b):
                    return left
                
                if(a > b):
                    return right
                
            return left
            
        
        def order(node, values = []):
            nonlocal res
            if(not node):
                return
            
            values = [node.val] + values
            print(values)
            order(node.left, values)
            order(node.right, values)
            if(not node.left and not node.right):
                # print("Entered Here", values)
                res = compare(values, res)
        
        
        
        res = [26]
        ans = []
        order(root)
        for v in res:
            ans.append(chr(97 + v))
        
        return ''.join(ans)
            
            
        

File: 0988-smallest-string-starting-from-leaf/0988-smallest-string-starting-from-leaf.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func smallestFromLeaf(root *TreeNode) string {
    var res string;
    smallestLeaf(root, "", &res)
    return res
}

func smallestLeaf(root *TreeNode, chars string, res *string){
    if root == nil{
        return
    }
    
    temp := string(97 + root.Val) + chars
    
    if root.Left == nil && root.Right == nil {
        if len(*res) == 0 {
            *res = temp
        } else {
           *res = smallestString(*res, temp) 
        }
        return
    }
    smallestLeaf(root.Left, temp, res)
    smallestLeaf(root.Right, temp, res)
}

func smallestString(str1, str2 string) string{
    for i, char := range str1 {
        if i >= len(str2) {
            return str2
        }
        
        if char < rune(str2[i]) {
            return str1
        } else if char > rune(str2[i]) {
            return str2
        }
    }
    
    if len(str1) < len(str2){
        return str1
    }
    return str2
}